Route module

bartverhaar opened the pull request:
ModuleRoute added. This module reads a route from a location on the viewer onboard and displays the full route on the map. For now only a route from Simacar is available.

bartverhaar committed:
Added route module for showing route from simacar in VRGZ

bartverhaar committed:
Fixed spaces

bartverhaar committed:
added check for response on routeUrl at register

matthijsln commented on public/js/safetymaps/modules/vrgz/route.js:
```diff
@@ -0,0 +1,131 @@
+    id: "dbkjs.modules.route",
+    options: null,
+    reloadTimeOut: null,
+    routeUrlSimacar: "/simacar_export/Route.txt",
```
Maybe put `routeUrlSimacar` in the default options so it can be overridden

matthijsln replied:
Do you have a test file with a route so I can test it? Maybe put it in a `test` directory next to `public` at the root of the repository.

matthijsln commented on public/js/safetymaps/modules/vrgz/route.js:
```diff
@@ -0,0 +1,131 @@
+        var me = this;
+
+        if (me.options.provider === "simacar")            
+            me.getSimacarRoutesAsPointArray(function (pointArray) {
```
Always use { }, also for single line statements

matthijsln replied:
Use jQuery promises instead of callbacks

matthijsln commented on public/js/safetymaps/modules/vrgz/route.js:
```diff
@@ -0,0 +1,131 @@
+
+        if (me.options.provider === "simacar")            
+            me.getSimacarRoutesAsPointArray(function (pointArray) {
+                me.addRouteLayerToMap(me.createRouteLayer(pointArray), reload);
```
Don't add layer everytime you get a new route, just do it once on startup and on update remove all features and only do `routeLayer.layer.addFeatures()`

matthijsln commented on public/js/safetymaps/modules/vrgz/route.js:
```diff
@@ -0,0 +1,131 @@
+            });
+
+        me.reloadTimeOut = window.setTimeout(function () {
+            if ((!reload && !me.routeUrlNotResponded) || reload)
```
Not very clear reload logic, maybe clearer to use jQuery promise always()

matthijsln reviewed (changes requested):
Looks like a cool feature!

matthijsln commented:
When viewing a viewer online instead of onboard this module will generate a 404 XHR every 5 seconds cluttering the network DevTools tab, I've made various ways to enable modules when running onboard only but they're all kind of a hack...

bartverhaar committed:
Refoctored. See comments from @matthijsln on PullRequest #23

bartverhaar commented:
> When viewing a viewer online instead of onboard this module will generate a 404 XHR every 5 seconds cluttering the network DevTools tab, I've made various ways to enable modules when running onboard only but they're all kind of a hack...

Added states to solve this.

matthijsln committed:
moved to modules, not customer specific
getRoute was Simacar specific, changed to getRouteFromProvider() using $.Deferred
added 'var' to undeclared variables
added missing break from case statement
simplified layer initialization
simplified feature building code
simplified vector layer styling
removed inactive state, replaced by longer retry timeout

matthijsln committed:
added button to zoom to route (or endpoint if to big)

matthijsln committed:
use Array filter() and map() functions

matthijsln committed:
route: enable ifModified cache for simacar Route.txt XHR

matthijsln committed:
added logging

matthijsln reviewed (approved):
Refactored more, added a button to zoom to the route and added caching.

matthijsln committed:
updateRoute() is not called when not modified, simplify
